pycolmap/custom_bundle_adjustment.py

```diff
@@ -0,0 +1,440 @@
+                ba_config.add_variable_point(point3D_id)
+                variable_point3D_ids.add(point3D_id)
+
+        # Adjust the local bundle
+        summary = solve_bundle_adjustment(
+            mapper.get_reconstruction(), ba_options, ba_config
+        )
+        logging.info("Local Bundle Adjustment")
+        logging.info(summary.BriefReport())
+
+        report.num_adjusted_observations = int(summary.num_residuals / 2)
+        # Merge refined tracks with other existing points
+        report.num_merged_observations = mapper.get_triangulator().merge_tracks(
+            tri_options, variable_point3D_ids
+        )
+        # Complete tracks that may have failed to triangulate before refinement
+        # of camera pose and calibration in bundle adjustment. This may avoid that some points are filtered and it helps for subsequent image registrations
+        report.num_completed_observations = (
+            mapper.get_triangulator().complete_tracks(
+                tri_options, variable_point3D_ids
+            )
+        )
+        report.num_completed_observations += (
+            mapper.get_triangulator().complete_image(tri_options, image_id)
+        )
+
+    filter_image_ids = {image_id, *local_bundle}
+    report.num_filtered_observations = reconstruction.filter_points3D_in_images(
+        mapper_options.filter_max_reproj_error,
+        mapper_options.filter_min_tri_angle,
+        filter_image_ids,
+    )
+    report.num_filtered_observations += reconstruction.filter_points3D(
+        mapper_options.filter_max_reproj_error,
+        mapper_options.filter_min_tri_angle,
+        point3D_ids,
+    )
+    return report
+
+
+def iterative_local_refinement(
+    mapper,
+    max_num_refinements,
+    max_refinement_change,
+    mapper_options,
+    ba_options,
+    tri_options,
+    image_id,
+):
+    """Equivalent to mapper.iterative_local_refinement(...)"""
+    ba_options_tmp = copy.deepcopy(ba_options)
+    for i in range(max_num_refinements):
+        # report = mapper.adjust_local_bundle(mapper_options, ba_options_tmp, tri_options, image_id, mapper.get_modified_points3D())
+        report = adjust_local_bundle(
+            mapper,
+            mapper_options,
+            ba_options_tmp,
+            tri_options,
+            image_id,
+            mapper.get_modified_points3D(),
+        )
+        logging.verbose(
+            1,
+            "=> Merged observations: {0}".format(
+                report.num_merged_observations
+            ),
+        )
+        logging.verbose(
+            1,
+            "=> Completed observations: {0}".format(
+                report.num_completed_observations
+            ),
+        )
+        logging.verbose(
+            1,
+            "=> Filtered observations: {0}".format(
+                report.num_filtered_observations
+            ),
+        )
+        changed = 0
+        if report.num_adjusted_observations != 0:
```

Comment:
```suggestion
        if report.num_adjusted_observations:
```

Reply:
However, in the guide it says

> "When handling integers, implicit false may involve more risk than benefit (i.e., accidentally handling None as 0). You may compare a value which is known to be an integer (and is not the result of len()) against the integer 0."

Should we still update this? I updated it to "> 0" for now

Reply:
No strong opinion, the explicit type check is indeed nice.